In [1]:
#import
import warnings
warnings.filterwarnings("ignore")
import matplotlib.pyplot as plt

import numpy as np
import pandas as pd

# from matplotlib import style
import seaborn as sns
%matplotlib inline

import acquire as a
import prepare as prep


# acquire

- we aquire a data set  from "anonymized-curriculum-access-07-2021.txt (df1)
- we acquire cohorts table from sql (df2)
- we merge both tables df1 and df2 using left join so we can keep all df1 observations and columns and add df2 columns
- before the merge we drop colums ['deleted_at' , 'slack']

In [2]:
df = a.acquire()

In [3]:
df.head()

date                             endpoint  user_id  cohort_id  \
0  2018-01-26                                    /        1        8.0   
1  2018-01-26                              java-ii        1        8.0   
2  2018-01-26  java-ii/object-oriented-programming        1        8.0   
3  2018-01-26   slides/object_oriented_programming        1        8.0   
4  2018-01-26            javascript-i/conditionals        2       22.0   

      source_ip     name  start_date    end_date           created_at  \
0  97.105.19.61  Hampton  2015-09-22  2016-02-06  2016-06-14 19:52:26   
1  97.105.19.61  Hampton  2015-09-22  2016-02-06  2016-06-14 19:52:26   
2  97.105.19.61  Hampton  2015-09-22  2016-02-06  2016-06-14 19:52:26   
3  97.105.19.61  Hampton  2015-09-22  2016-02-06  2016-06-14 19:52:26   
4  97.105.19.61    Teddy  2018-01-08  2018-05-17  2018-01-08 13:59:10   

            updated_at  program_id  
0  2016-06-14 19:52:26         1.0  
1  2016-06-14 19:52:26         1.0  
2  2016-06-14 19:52:26         1.0  
3  2016-06-14 19:52:26         1.0  
4  2018-01-08 13:59:10         2.0

In [4]:
a.miss_dup_values(df)

Your selected dataframe has 11 columns.
There are 8 columns that have missing values.
  
** There are 394392 duplicate rows that represents 38.71% of total Values**


Missing Values  % of Total Values
name                 64497                6.3
start_date           64497                6.3
end_date             64497                6.3
created_at           64497                6.3
updated_at           64497                6.3
program_id           64497                6.3
cohort_id            53497                5.3
endpoint                 1                0.0

# Prepare

In [5]:
#use my functions
df = prep.prepare(df)

In [6]:
a.miss_dup_values(df)

Your selected dataframe has 10 columns.
There are 4 columns that have missing values.
  
** There are 805477 duplicate rows that represents 79.06% of total Values**


Missing Values  % of Total Values
created_at           64497                6.3
updated_at           64497                6.3
cohort_id            53497                5.3
name                 53497                5.3

In [8]:
df.cohort_id.value_counts().sort_index()

1.0      19063
2.0         95
4.0          4
5.0          1
6.0         72
7.0        833
8.0       1712
9.0          5
11.0       253
12.0       302
13.0      2848
14.0      9596
15.0       691
16.0       755
17.0      4960
18.0      2257
19.0      1371
21.0      7444
22.0     31258
23.0     28581
24.0     35707
25.0     25680
26.0     27811
27.0     20864
28.0     96954
29.0     38198
31.0     25372
32.0     29754
33.0     40858
34.0     27261
51.0     35381
52.0     28177
53.0     37043
55.0     22063
56.0     34172
57.0     34919
58.0     30137
59.0     32871
61.0     18212
62.0     39316
132.0    24978
133.0    18371
134.0    21095
135.0    34093
137.0    21435
138.0    23744
139.0    17745
165.0     9171
166.0     1829
Name: cohort_id, dtype: int64

**prepare**
- we find 165 and 166 cohorts_id whithout name so we assigned a new name
- we fill na with zero ['start_date', 'end_date', 'program_id']
- change start_date, end_date and date change to date type
- set date to index
- drop a row with null value in endpoint column
    

**1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?**

   **-lesson that attract the most traffic consistently across cohorts**

In [ ]:
df.groupby(df. )